Develop connect api

Ja7423 opened the pull request:
目前只完成取得device token而已
還不是很確定後台要什麼資料

Ja7423 committed:
feat: 完成連線到apns取得device token

happiness9721 commented on Pigeon/Classes/PigeonService.swift:
```diff
@@ -14,23 +14,45 @@ public class PigeonService: NSObject {
   private(set) static var appKey: String?
   private(set) static var deviceToken: String?
 
+  private static var pigeonService: PigeonService?
```
為什麼需要開一個 static 的 PegeonService?

Ja7423 replied:
我想不到UNUserNotificationCenter.current().delegate要asign給誰

Ja7423 replied:
有什麼好的做法嗎？

Ja7423 committed:
修改PigeonService為singleton
feat: 新增APIService，串接api
初步完成註冊手機流程，細節再補上

Ja7423 committed:
1. feat: 新增KeychainService儲存pigeonToken
2. regist device前先判斷有沒有token已經存在

Ja7423 committed:
1.keychain service 新增account屬性
2.patch body攜帶device token

happiness9721 commented on Pigeon/Classes/Models/Device.swift:
```diff
@@ -11,7 +11,16 @@ public struct Device: Codable {
+}
 
+extension Device {
+  func body() -> [String : Any] {
```
coding style 有問題，swiftlint的warning先修掉

happiness9721 commented on Pigeon/Classes/Models/Device.swift:
```diff
@@ -11,7 +11,16 @@ public struct Device: Codable {
+extension Device {
+  func body() -> [String : Any] {
+    return ["device_token": deviceToken ?? "",
+               "device_model": deviceModel ?? "",
```
這段重新排版一下

happiness9721 commented on Pigeon/Classes/PigeonService.swift:
```diff
@@ -38,10 +84,46 @@ public class PigeonService: NSObject {
+    }
   }
 
+  private func deviceFrom(_ data: Data) -> Device {
```
依照swift的命名習慣，改成`device(from: data: Data) -> Device`
會比較好

happiness9721 committed:
chore: Configure siwft lint.

happiness9721 committed:
chore: Update Nimble.

Ja7423 committed:
chore: fix lint warning

arthurgau0419 commented on Pigeon/Classes/APIService.swift:
```diff
@@ -0,0 +1,79 @@
+    request.httpMethod = HttpMethod.POST.rawValue
+    request.httpBody = body
+    request.allHTTPHeaderFields = ["app_key": appKey]
+
```
關於 http header
如果request body是json，加入 Content-Type: application/json
如果希望success response、error response是json，加入：Accept: application/json
每個Request在發出之前都要加

arthurgau0419 commented on Pigeon/Classes/KeychainService.swift:
```diff
@@ -0,0 +1,76 @@
+  case noToken
+  case unexpectedTokenData
+  case unexpectedItemData
+  case unhandledError(status: OSStatus)
```
在Xcode 10好像會報錯
Identifier Name Violation: Enum element name should only contain alphanumeric characters: 'unhandledError(status:)' (identifier_name)

happiness9721 replied:
這個是swiftlint的問題，更新就好了

arthurgau0419 reviewed (approved):
有些小地方以後可能會遇到問題，可以先改一下

Ja7423 committed:
chores: request的header加上content-type
        public function補上註解

happiness9721 commented on Pigeon/Classes/APIService.swift:
```diff
@@ -0,0 +1,88 @@
+
+  private func performTask(_ request: URLRequest, completionHandler: @escaping (Data) -> Void, errorHandler: @escaping (Error) -> Void) {
+    URLSession.shared.dataTask(with: request) { (data, response, error) in
+      print("---------registerDevice--------------")
```
可以用print的方式作debug，但是不要將debug用的code push上來

happiness9721 commented on Pigeon/Classes/APIService.swift:
```diff
@@ -0,0 +1,88 @@
+        errorHandler(ServiceError.noResponse)
+      }
+
+      }.resume()
```
縮排

happiness9721 commented on Pigeon/Classes/APIService.swift:
```diff
@@ -0,0 +1,88 @@
+      if let err = error {
+        errorHandler(err)
+      } else if let resp = response {
+        if self.validateStatusCode(resp),
```
這邊太多層if else了，改成guard的寫法

happiness9721 commented on Pigeon/Classes/PigeonService.swift:
```diff
@@ -8,27 +8,84 @@
     }
-    fatalError("Must Implement")
+
+    print("deviceToken : \(self.deviceToken)")
```
Xcode warning
print的內容不能為optional

happiness9721 commented on Pigeon/Classes/PigeonService.swift:
```diff
@@ -8,27 +8,84 @@
+    register(device: device, onCompleted: { [unowned self] (device) in
+      self.savePigeonToken(device)
+    }) { (error) in
+      print("register error: \(error)")
```
Xcode warning
print的內容不能為optional

happiness9721 commented on Pigeon/Classes/APIService.swift:
```diff
@@ -0,0 +1,88 @@
+
+    performTask(request, completionHandler: { (data) in
+      completionHandler(data)
+    }) { (error) in
```
swiftlint錯誤，參數超過兩個closure的時候，不要用這種省略參數名稱的寫法

Ja7423 committed:
chore: 改進review提出的錯誤
       每個request加上Accept json type
feat: 加入PigeonServiceError處理exception
      PigeonRegisterDelegate新增拋出error的function

Ja7423 committed:
feat:新增Validation，驗證status code
     error改成使用associated values的方式回傳

Ja7423 committed:
修改PigeonServiceError對應不同的error，回傳不同的detail information
